### PR TITLE
[REFACTOR] Use more TypedPackedFuncs

### PR DESCRIPTION
**Pull request by kazimuth**
This PR converts a good chunk of TVM's global `PackedFunc`s to be `TypedPackedFunc`s. Right now, this shouldn't actually change any behavior, since the type information is immediately discarded once the functions are registered in the `Registry`. 

It does make things a lot shorter, using a few helper methods I added on `Registry`: `set_body_simple`, `set_body_method`, and `set_body_node_method` (names up for debate). These helpers use the properties of function pointers to infer argument and return types:

```cpp
TVM_REGISTER_API("make.Realize")
.set_body([](TVMArgs args,  TVMRetValue *ret) {
    *ret = Realize::make(args[0], args[1], args[2], args[3], args[4], args[5]);
  }); 
```

becomes ->

```
TVM_REGISTER_API("make.Realize")
.set_body_simple(Realize::make);
```

This is shorter, runtime-equivalent, and also now has type information!

I don't believe this change has much runtime cost. It might interfere with inlining in some cases, would have to measure. Also, `PackedFunc`s aren't intended to be all that fast anyway.

So, that's the refactoring part. Now, the RFC part:

(moved to #2983)

### CODE REVIEW

**Comment by kazimuth** [comment]
cc @nhynes @tqchen

File: src/runtime/rpc/rpc_event_impl.cc [hunk 1]
```diff
@@ -44,9 +44,9 @@ PackedFunc CreateEventDrivenServer(PackedFunc fsend,
     });
 }
 
+std::function<PackedFunc(PackedFunc, std::string, std::string)> q = CreateEventDrivenServer;
```
Comment (kazimuth) [inline src/runtime/rpc/rpc_event_impl.cc]: whoops, this was for debugging, i should remove it

**Comment by tqchen** [comment]
Thanks for the contribution, I like the automated detection of the signatures. The main question is the naming convention to make sure we are not confusing the users. Let us open a new issue about other parts of RFC. 

I will summarize some of my take here(which can be moved to the RFC).  I like the idea of Node hierarchy compile time generation. This is something we have thought about for a while and might help https://github.com/dmlc/tvm/issues/2523#issuecomment-458821056

It is always tempting to automate the wrapper generation. However, our past experience teaches me that the automatic wrapper generation is never perfect. Think about how can we support keyword arguments, good pythonic style docstring and so on.  Eventually, we find that it is simpler to just do a manual wrapping, which gives us all the good native features, docs, and keep PackedFunc simple (by only support positional arguments without any meta-data).

**Comment by nhynes** [comment]
This idea actually comes a lot :P https://github.com/dmlc/tvm/pull/2328#issuecomment-450001679

I know, for sure, that we could get good docs with _really good_ codegen like that offered by Rust macros, but I also know for sure that we're not about to rewrite TVM in Rust :)

I think that the boilerplate really does bother new (advanced) users who want to use TVM as a tool. I wonder if there's a way forward here that satisfies all desiderata?

**Comment by kazimuth** [comment]
> The main question is the naming convention to make sure we are not confusing the users.

Yeah I didn't put any thought into these, would welcome suggestions. Could do something like `set_body_to_function_pointer` / `set_body_to_method_pointer`, although those are a bit verbose.

**Comment by kazimuth** [comment]
This is ready to merge once we bikeshed the helper names

**Comment by tqchen** [comment]
Here are some thoughts about naming:

- set_body_simple -> set_body_typed, since it is really a method for automatic type detection
   - I did a quick check and seems overloading do work
- set_body_method, set_body_node_method-> set_body_method
   - Use is_base_of to detect whether a class is subclass of NodeRef or Node, and use a different specialization path.

We also need to update the comment a bit to reflect the template arguments

**Comment by kazimuth** [comment]
Done @tqchen

**Comment by kazimuth** [comment]
Uh, I think that last build just timed out? is there a way to re-run it?

**Comment by tqchen** [comment]
Thanks @kazimuth @nhynes , this is now merge !

**Pull request merged**